src/irmin-test/common.ml

```diff
@@ -100,18 +100,9 @@ module Suite = struct
       | S (module S) -> (module S : Generic_key)
     in
     let open Lwt.Syntax in
-    let module B = Store.Backend in
-    let clear repo =
-      Lwt.join
-        [
-          B.Commit.clear (B.Repo.commit_t repo);
-          B.Node.clear (B.Repo.node_t repo);
-          B.Contents.clear (B.Repo.contents_t repo);
-          B.Branch.clear (B.Repo.branch_t repo);
-        ]
-    in
     let* repo = Store.Repo.v config in
-    let* () = clear repo in
```

Comment:
I spent a while looking into this, and think this removed `clear` is the reason the CI is failing. It's a bit subtle. The [`test_wide_nodes` test](https://github.com/mirage/irmin/blob/e54478bd4b800ad83eeb7ad93632555308fad6f7/src/irmin-test/store.ml#L1807-L1813) builds a large tree, exports it, and then expects the tree to be convertible to a backend node.

This is fine provided that the export actually happened (since the tree will then be a key into the backend's node store), but _not_ fine if the export was skipped due to the tree's hash already existing in the backend [here](https://github.com/mirage/irmin/blob/e54478bd4b800ad83eeb7ad93632555308fad6f7/src/irmin/tree.ml#L2114-L2115). When we skip the node, we do actually cache the key that witnessed membership [here](https://github.com/mirage/irmin/blob/e54478bd4b800ad83eeb7ad93632555308fad6f7/src/irmin/tree.ml#L2011), but this has no effect since we can't use this key without also caching the `repo` (we'll miss [here](https://github.com/mirage/irmin/blob/e54478bd4b800ad83eeb7ad93632555308fad6f7/src/irmin/tree.ml#L698) in the case `Map _, Key _`). The result is that we can't dereference the backend node and fail [here](https://github.com/mirage/irmin/blob/e54478bd4b800ad83eeb7ad93632555308fad6f7/src/irmin/tree.ml#L979-L983) later in the test.

This skipped export happens because a prior test within `irmin-test` already exports this node: [`test_commit_wide_node`](https://github.com/mirage/irmin/blob/e54478bd4b800ad83eeb7ad93632555308fad6f7/src/irmin-test/store.ml#L1867). With the removal of this `clear` call here between the test cases, `test_commit_wide_node` causes `test_wide_nodes` to pick up a non-empty repository and fail (due to memoisation of `Store.Repo.v`). Running just `test_wide_nodes` by itself passes.

I think there's two problems here:

- I think `Tree.export repo t` should have the same impact on `t` regardless of whether the underlying tree already exists in `repo`: in particular, it should drop the in-memory `Map` and use a `Key` representation that retains the `repo` pointer. This ensures the invariant "_`Tree.export repo t` ensures `Tree.to_backend_node t` will succeed_", which the tests expect to hold & seems reasonable to me.

- more generally, without `clear`, we need a better way to keep the test cases in `irmin-test` independent of each other or else we'll keep running into such problems.

Reply:
Thanks for this details analysis.

The following patch seen to make the bug go away, but I'm not understanding why it works. 

```diff
diff --git a/src/irmin/tree.ml b/src/irmin/tree.ml
index 93e78ab1d7..b162588d91 100644
--- a/src/irmin/tree.ml
+++ b/src/irmin/tree.ml
@@ -810,8 +810,6 @@ module Make (P : Backend.S) = struct
       | Key (_, k) -> Some k
       | Map _ | Value _ | Portable_dirty _ | Pruned _ -> None

-    let unsafe_cache_key t k = t.info.ptr <- Key k
-
     (* When computing hashes of nodes, we try to use [P.Node.Val.t] as a
        pre-image if possible so that this intermediate value can be cached
        within [t.info.value] (in case it is about to be written to the backend).
@@ -2008,9 +2006,12 @@ module Make (P : Backend.S) = struct
                   (* We always prefer caching keys over hashes. In this case,
                      caching the key ensures that any yet-to-be-exported parent
                      nodes can reference this value. *)
-                  Node.unsafe_cache_key n k;
                   cnt.node_mem <- cnt.node_mem + 1;
-                  P.Node.mem node_t k))
+                  let+ mem = P.Node.mem node_t k in
+                  if mem then (
+                    Node.export ?clear repo n k;
+                    true)
+                  else false))
     in

     let add_node n v k =
```
---


>  I think `Tree.export repo t` should have the same impact on `t` regardless of whether the underlying tree already exists in `repo`

I agree and I think that my above patch is not enough to ensure that

Reply:
Thanks this is extremely helpful!